src/api/HttpApi.js

```diff
@@ -0,0 +1,284 @@
+/* @flow */
+/* eslint-disable no-underscore-dangle */
+
+import url from 'url';
+import DataLoader from 'dataloader';
+import { connectionFromArray, forwardConnectionArgs } from 'graphql-relay';
+import omit from 'lodash/omit';
+import pick from 'lodash/pick';
+import snakeCase from 'lodash/snakeCase';
+import querystring from 'querystring';
+
+import HttpError from './utils/HttpError';
+import request from './utils/request';
+import type { Data, Request } from './utils/request';
+
+const PAGINATION_ARG_KEYS = Object.keys(forwardConnectionArgs);
+
+export type Args = { [key: string]: mixed };
+export type { Data };
+
+export type ValidationResult = {
+  valid: boolean,
+  errors: ?Array<mixed>,
+};
+
+export type HttpApiOptions = {
+  authorization?: string,
+  getHeaders: Request => { [string]: string },
+  apiBase: string,
+  origin: string,
+  externalOrigin: string,
+};
+
+export default class HttpApi {
+  authorization: string;
+  _getHeaders: Request => { [string]: string };
+  _request: Request;
+  _origin: string;
+  _apiBase: string;
+  _externalOrigin: string;
+  _loader: DataLoader<*, *>;
+
+  constructor(
+    req: Request,
+    {
+      authorization,
+      apiBase,
+      getHeaders,
+      origin,
+      externalOrigin,
+    }: HttpApiOptions,
+  ) {
+    this.authorization = authorization || '';
+
+    this._getHeaders = getHeaders;
+    this._origin = origin;
+    this._externalOrigin = externalOrigin;
+    this._apiBase = apiBase;
+
+    this._loader = new DataLoader(paths =>
+      Promise.all(
+        paths.map(path =>
+          // Don't fail the entire batch on a single failed request.
+          this._get(path).catch(e => e),
+        ),
+      ),
+    );
+  }
+
+  async get<T>(path: string, args?: Args): Promise<?T> {
+    const item = await this._loader.load(this.makePath(path, args));
+    if (item instanceof Error) {
+      throw item;
+    }
+
+    return item;
+  }
+
+  async getPaginatedConnection(
+    path: string,
+    { after, first, ...args }: Args,
+  ): Promise<mixed> {
+    const items = await this.get(
+      this.makePath(path, {
+        ...args,
+        cursor: after,
+        limit: first,
+        pageSize: first,
+      }),
+    );
+    if (!items) {
+      return null;
+    }
+
+    // XXX Remove items.links when all endpoints use modern pagination
+    if (items.meta) {
+      // new style pagination
+      const { cursors, hasNextPage } = items.meta;
+
+      // These connections only paginate forward, so the existence of a prev
+      // page doesn't make any difference, but this is the correct value.
+      const hasPreviousPage = !!after;
+
+      return {
+        edges: items.map((item, i) => ({
+          node: item,
+          cursor: cursors[i],
+        })),
+        pageInfo: {
+          hasNextPage,
+          endCursor: cursors[cursors.length - 1],
+          hasPreviousPage,
+        },
+      };
+    } else if (items.links) {
```

Comment:
We might be able to drop this. I'm migrating over what I think is the last list that uses this.